### PR TITLE
Implement selectors

### PR DESCRIPTION
**Pull request by kenzierocks**
See https://github.com/SpongePowered/Sponge/pull/243.

Depends on API PR https://github.com/SpongePowered/SpongeAPI/pull/577.

### CODE REVIEW

**Comment by kenzierocks** [comment]
Continuation of #4.

**Comment by kenzierocks** [comment]
Ok, AFAIK this is good.

**Comment by gabizou** [comment]
@kenzierocks Did you test?

**Comment by kenzierocks** [comment]
Limited testing, I'll run some extensive ones ASAP.

**Comment by kenzierocks** [comment]
...apparently it didn't apply an AT when I rebased, let me patch that.

**Comment by kenzierocks** [comment]
Alright, should be good, if anyone wants something specific tested I can test that.

File: src/main/java/org/spongepowered/common/registry/SpongeGameRegistry.java [hunk 1]
```diff
@@ -26,11 +26,22 @@
 
 import static com.google.common.base.Preconditions.checkNotNull;
 
-import com.google.common.base.Function;
-import com.google.common.collect.ImmutableBiMap;
-import com.google.common.collect.ImmutableList;
-import com.google.common.collect.ImmutableMap;
-import com.google.common.collect.Maps;
+import java.awt.Color;
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/registry/SpongeGameRegistry.java]: Wrong import order

File: src/main/java/org/spongepowered/common/registry/SpongeGameRegistry.java [hunk 2]
```diff
@@ -448,7 +462,7 @@
                     .put(RailDirection.class, ImmutableMap.<String, CatalogType>of()) // TODO
                     .put(Rotation.class, ImmutableMap.<String, CatalogType>of()) // TODO
                     .put(SandstoneType.class, ImmutableMap.<String, CatalogType>of()) // TODO
-                    .put(SelectorType.class, ImmutableMap.<String, CatalogType>of()) // TODO
+                    .put(SelectorType.class, this.selectorMappings) // TODO
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/registry/SpongeGameRegistry.java]: Does the TODO comment still apply?

Comment (kenzierocks) [inline src/main/java/org/spongepowered/common/registry/SpongeGameRegistry.java]: No, must have missed it.

File: src/main/java/org/spongepowered/common/text/selector/SpongeSelector.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    @Override
+    public String toPlain() {
+        return this.plain;
+    }
+
+    @Override
+    public SelectorBuilder builder() {
+        return null;
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/text/selector/SpongeSelector.java]: Need TODO comment and could throw `UnsupportedOperationException` so this won't go unnoticed.

Comment (kenzierocks) [inline src/main/java/org/spongepowered/common/text/selector/SpongeSelector.java]: Oh. This is what happens when I don't test builders. I'll implement this.

Comment (Minecrell) [inline src/main/java/org/spongepowered/common/text/selector/SpongeSelector.java]: Might be also a good idea to test them ;P

**Comment by kenzierocks** [comment]
Tested more, everything works, I'll work on the merge conflicts.

**Comment by kenzierocks** [comment]
And good.

**Comment by zml2008** [comment]
Has this been updated for the latest has()/isInverted() changes to the API?

**Comment by kenzierocks** [comment]
Locally, yes, but I'm still implementing extent support.

**Comment by kenzierocks** [comment]
Alright I think it's good. Currently impl acts like vanilla for resolving wherever it was applicable.

File: src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java [hunk 4]
```diff
@@ -215,7 +215,7 @@ public String getIdentifier() {
 
     @Override
     public Tristate permDefault(String permission) {
-        return Tristate.FALSE;
+        return Tristate.TRUE; // temporary patch for OP status
```
Comment (zml2008) [inline src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java]: no

Comment (kenzierocks) [inline src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java]: Whoops, that's a local change I didn't mean to push.

File: src/main/java/org/spongepowered/common/text/selector/SelectorResolver.java [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+        builder.addAll(ArgumentTypes.POSITION.getTypes());
+        builder.addAll(ArgumentTypes.DIMENSION.getTypes());
+        builder.addAll(ArgumentTypes.RADIUS.getTypes());
+        // Left commented because Vanilla doesn't include it (see field_179666_d)
+        // builder.addAll(ArgumentTypes.ROTATION.getTypes());
+        LOCATION_BASED_ARGUMENTS = builder.build();
+    }
+
+    private static Extent extentFromSource(CommandSource origin) {
+        if (origin instanceof ICommandSender) {
```
Comment (zml2008) [inline src/main/java/org/spongepowered/common/text/selector/SelectorResolver.java]: Use LocatedSource here instead

File: src/main/java/org/spongepowered/common/text/selector/SelectorResolver.java [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+            return (Extent) ((ICommandSender) origin).getEntityWorld();
+        }
+        return null;
+    }
+
+    private static Vec3 positionFromSource(CommandSource origin) {
+        if (origin instanceof ICommandSender) {
```
Comment (zml2008) [inline src/main/java/org/spongepowered/common/text/selector/SelectorResolver.java]: Use LocatedSource here instead

**Pull request closed without merging**